src/core/consumerthread.h

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CORE_CONSUMERTHREAD_H
+#define BITCOIN_CORE_CONSUMERTHREAD_H
+
+#include <future>
+#include <thread>
+
+#include <core/producerconsumerqueue.h>
+#include <util.h>
+
+template <WorkerMode MODE>
+class ConsumerThread;
+
+//! A WorkItem() encapsulates a task that can be processed by a ConsumerThread()
+//! @see ConsumerThread()
+template <WorkerMode MODE>
+class WorkItem
+{
+    friend ConsumerThread<MODE>; //<! only a consumer thread can execute a WorkItem
+
+protected:
+    WorkItem(){};
+    virtual void operator()(){};
+};
+
+template <WorkerMode MODE>
+class GenericWorkItem : public WorkItem<MODE>
+{
+    friend ConsumerThread<MODE>;
+
+public:
+    GenericWorkItem(std::function<void()> f) : m_f(f) {}
+
+protected:
+    void operator()() override
+    {
+        m_f();
+    }
+
+    std::function<void()> m_f;
+};
+
+//! A special WorkItem() that is used to interrupt a blocked ConsumerThread() so that it can terminate
+template <WorkerMode MODE>
+class ShutdownPill : public WorkItem<MODE>
+{
+    friend ConsumerThread<MODE>;
+
+private:
+    ShutdownPill(ConsumerThread<MODE>& consumer) : m_consumer(consumer){};
+    void operator()()
+    {
+        std::thread::id id = m_consumer.m_thread.get_id();
+        if (std::this_thread::get_id() != id) {
+            // this ShutdownPill was intended for another thread
+
+            // we haven't seen this pill before
+            if (!m_threads_observed.count(id)) {
+                m_threads_observed.insert(std::this_thread::get_id());
+
+                // resubmit it so that it gets a chance to get to the right thread
+                // when resubmitting, do not block and do not care about failures
+                // theres a potential deadlock where we try to push this to a queue thats
+                // full and there are no other threads still consuming
+                // since the only purpose of reinjecting this is to terminate threads that
+                // may be blocking on an empty queue when the queue is full we do not need to do this
+                m_consumer.m_queue->Push(MakeUnique<ShutdownPill<MODE>>(std::move(*this)), WorkerMode::NONBLOCKING);
+            }
+
+            // if the same pill has been seen by the same thread previously then it can safely be discarded
+            // the intended thread has either terminated or is currently processing a work item and will terminate
+            // after completing that item and before blocking on the queue
```

Comment:
In commit "Add ConsumerThread: to consumer and operate on work from a ProducerConsumerQueue" (7f8a888)

Why couldn't the intended thread just be blocked calling Pop() and not terminated or currently processing anything? It seems like this is assuming threads are notified in a circular order.